src/datadog/tracer_config.cpp

```diff
@@ -119,6 +122,134 @@ Expected<std::unordered_map<std::string, std::string>> parse_tags(
   return tags;
 }
 
+// Return a `PropagationStyles` parsed from the specified `env_var`.
+// If `env_var` is not in the environment, return `nullopt`.
+// If an error occurs, throw an `Error`.
+Optional<PropagationStyles> styles_from_env(environment::Variable env_var) {
+  const auto styles_env = lookup(env_var);
+  if (!styles_env) {
+    return {};
+  }
+
+  auto styles = parse_propagation_styles(*styles_env);
+  if (auto *error = styles.if_error()) {
+    std::string prefix;
+    prefix += "Unable to parse ";
+    append(prefix, name(env_var));
+    prefix += " environment variable: ";
+    throw error->with_prefix(prefix);
+  }
+  return *styles;
+}
+
+std::string json_quoted(StringView text) {
+  std::string unquoted;
+  assign(unquoted, text);
+  return nlohmann::json(unquoted).dump();
+}
+
+Expected<void> finalize_propagation_styles(FinalizedTracerConfig &result,
+                                           const TracerConfig &config,
+                                           Logger &logger) {
+  namespace env = environment;
+  // Print a warning if a questionable combination of environment variables is
+  // defined.
+  const auto ts = env::DD_TRACE_PROPAGATION_STYLE;
+  const auto tse = env::DD_TRACE_PROPAGATION_STYLE_EXTRACT;
+  const auto se = env::DD_PROPAGATION_STYLE_EXTRACT;
+  const auto tsi = env::DD_TRACE_PROPAGATION_STYLE_INJECT;
+  const auto si = env::DD_PROPAGATION_STYLE_INJECT;
+  // clang-format off
+  /*
+           ts    tse   se    tsi   si
+           ---   ---   ---   ---   ---
+    ts  |  x     warn  warn  warn  warn
+        |
+    tse |  x     x     warn  ok    ok
+        |
+    se  |  x     x     x     ok    ok
+        |
+    tsi |  x     x     x     x     warn
+        |
+    si  |  x     x     x     x     x
+  */
+  // In each pair, the first would be overridden by the second.
+  const std::pair<env::Variable, env::Variable> questionable_combinations[] = {
+           {ts, tse}, {ts, se},  {ts, tsi}, {ts, si},
+
+                      {se, tse}, /* ok */   /* ok */
```

Comment:
Did formatting ruin the comments here a bit?

Reply:
They're meant to reflect the diagram commented above. I used vim for fixed with fonts :)

Maybe something was lost in translation, I'll check...